[3.x] Fix top level `CanvasItem` visibility

timothyqiu opened the pull request:
This PR first reverts the backported #57684 (part of #57900) and #58386.

#57684 was backported mainly because it solves a bug where

> CanvasLayer could propagate its visibility to CanvasItems that were separated by plain Nodes.

But after inspection these days, I think that is the expected behavior. `CanvasItem`s that are explicitly marked as top level or have a non-`CanvasItem` parent were handled together as "top level", and they are logically parented to the nearest ancestor `CanvasLayer` or `Viewport` world canvas. See [`CanvasItem::_enter_canvas()`](https://github.com/godotengine/godot/blob/6b4d7d20a48ddcc5bf457df38053086ab6041c9f/scene/2d/canvas_item.cpp#L498-L534).

After the revert, the only problem is that `CanvasLayer` visibility propagation was only propagated to direct child `CanvasItem`s. Fortunately, all these "top level" nodes are in a group called `root_canvas{canvas_rid}`, so moving the propagation logic into a `call_group()` is all that have to be done I think.

The editor gizmo fix from the reverted 642591b6a96285d70cd1fefc6b7f997a1395c07f is kept.

---

I also stopped propagation onto explicitly top level `CanvasItem`s because:

* Their logical parent is a `CanvasLayer`.
* For `CanvasLayer`, we propagate to these `CanvasItem`s with `call_group()`.
* There was previously a comment:
    https://github.com/godotengine/godot/blob/6b4d7d20a48ddcc5bf457df38053086ab6041c9f/scene/2d/canvas_item.cpp#L385-L387
* Previously, the visibility of `CanvasItem`s are only controlled by their logical parent inside `VisualServer`. So propagating onto top level nodes won't actually change anything.

---

Tested recent regressions like #58251, #58315, and #58782. They are not reproducible.

Fixes #58782

akien-mga commented:
It would be clearer for the Git history if you could do this as actually `git revert` commits + another commit on top with your further changes.

timothyqiu committed:
Revert "Fix CanvasItem visibility propagation"

This reverts commit 642591b6a96285d70cd1fefc6b7f997a1395c07f.

timothyqiu commented:
> It would be clearer for the Git history if you could do this as actually `git revert` commits + another commit on top with your further changes.

The backport of #57684 is part of #57900. It can only be reverted manually. Should I do that as a separate commit too?

timothyqiu committed:
Revert backport of "Rework CanvasItem visibility propagation"

This reverts part of commit f49ffe4bb02c6af443652f67bbf6c0c47344d721

timothyqiu committed:
Fix top level CanvasItem visibility

The editor gizmo fix from previously reverted
642591b6a96285d70cd1fefc6b7f997a1395c07f is kept here.

akien-mga commented:
I meant you could revert this commit no? https://github.com/godotengine/godot/commit/f49ffe4bb02c6af443652f67bbf6c0c47344d721

akien-mga commented:
Thanks!

timothyqiu commented:
> I meant you could revert this commit no? [f49ffe4](https://github.com/godotengine/godot/commit/f49ffe4bb02c6af443652f67bbf6c0c47344d721)

#57900 backports 3 PRs in that one commit f49ffe4bb02c6af443652f67bbf6c0c47344d721. It's the last one of the three that needs revert. `CanvasLayer` visibility is kept.

I think I should have made that backport PR 3 separate commits in the first place :(
